File: src/domain/outfit-filters.test.ts
import { describe, expect, it } from 'vitest';
import {
  NO_FILTERS,
  isUnfiltered,
  occasionChips,
  seasonChips,
  withOccasionSelected,
  withSeasonToggled,
} from './outfit-filters';

const activeSeasons = (filters = NO_FILTERS) =>
  seasonChips(filters).filter(c => c.active).map(c => c.value);

const activeOccasions = (filters = NO_FILTERS) =>
  occasionChips(filters).filter(c => c.active).map(c => c.value);

describe('the season row', () => {
  it('shows "any" as the only thing on when nothing is chosen', () => {
    expect(activeSeasons()).toEqual([null]);
  });

  it('turns "any" off as soon as a season is chosen', () => {
    const filters = withSeasonToggled(NO_FILTERS, 'summer');

    expect(activeSeasons(filters)).toEqual(['summer']);
  });

  it('holds several seasons at once', () => {
    // A garment for spring is often a garment for fall, so this is a set.
    let filters = withSeasonToggled(NO_FILTERS, 'spring');
    filters = withSeasonToggled(filters, 'fall');

    expect(activeSeasons(filters)).toEqual(['spring', 'fall']);
  });

  it('takes a season off when it is tapped again', () => {
    let filters = withSeasonToggled(NO_FILTERS, 'summer');
    filters = withSeasonToggled(filters, 'summer');

    expect(activeSeasons(filters)).toEqual([null]);
    expect(isUnfiltered(filters)).toBe(true);
  });

  it('lands in the same state whether emptied by tapping or by "any"', () => {
    // "any" is derived rather than stored, so there is only one empty state to
    // reach -- which is what makes the row always readable one way.
    const tappedOff = withSeasonToggled(withSeasonToggled(NO_FILTERS, 'winter'), 'winter');
    const cleared = withSeasonToggled(withSeasonToggled(NO_FILTERS, 'winter'), null);

    expect(tappedOff).toEqual(cleared);
  });

  it('clears everything when "any" is tapped', () => {
    let filters = withSeasonToggled(NO_FILTERS, 'spring');
    filters = withSeasonToggled(filters, 'summer');
    filters = withSeasonToggled(filters, null);

    expect(activeSeasons(filters)).toEqual([null]);
  });

  it('keeps seasons in the app order, not the order they were tapped', () => {
    // So the chips and the stored state agree, and so the same choice always
    // reads the same. The engine is indifferent -- it tests membership -- which
    // is what makes this safe to normalise.
    let filters = withSeasonToggled(NO_FILTERS, 'winter');
    filters = withSeasonToggled(filters, 'spring');

    expect(filters.seasons).toEqual(['spring', 'winter']);
  });

  it('offers "any" plus every season, once each', () => {
    expect(seasonChips(NO_FILTERS).map(c => c.value)).toEqual([
      null, 'spring', 'summer', 'fall', 'winter', 'all-season',
    ]);
  });
});

describe('the occasion row', () => {
  it('shows "any" as the only thing on when nothing is chosen', () => {
    expect(activeOccasions()).toEqual([null]);
  });

  it('holds one choice at a time', () => {
    // An outfit is for one thing, unlike a season.
    let filters = withOccasionSelected(NO_FILTERS, 'work');
    filters = withOccasionSelected(filters, 'sport');

    expect(activeOccasions(filters)).toEqual(['sport']);
  });

  it('clears the choice when the active chip is tapped again', () => {
    let filters = withOccasionSelected(NO_FILTERS, 'formal');
    filters = withOccasionSelected(filters, 'formal');

    expect(activeOccasions(filters)).toEqual([null]);
  });

  it('lands in the same state whether cleared by re-tapping or by "any"', () => {
    const reTapped = withOccasionSelected(withOccasionSelected(NO_FILTERS, 'lounge'), 'lounge');
    const cleared = withOccasionSelected(withOccasionSelected(NO_FILTERS, 'lounge'), null);

    expect(reTapped).toEqual(cleared);
  });

  it('offers "any" plus every occasion, once each', () => {
    expect(occasionChips(NO_FILTERS).map(c => c.value)).toEqual([
      null, 'casual', 'work', 'formal', 'sport', 'lounge',
    ]);
  });
});

describe('the two rows together', () => {
  it('keeps a season choice when the occasion changes, and the other way round', () => {
    let filters = withSeasonToggled(NO_FILTERS, 'summer');
    filters = withOccasionSelected(filters, 'work');

    expect(filters).toEqual({ seasons: ['summer'], occasion: 'work' });
    expect(isUnfiltered(filters)).toBe(false);

    filters = withSeasonToggled(filters, null);
    expect(filters).toEqual({ seasons: [], occasion: 'work' });
    expect(isUnfiltered(filters)).toBe(false);
  });

  it('is unfiltered only when both rows are empty', () => {
    expect(isUnfiltered(NO_FILTERS)).toBe(true);
    expect(isUnfiltered(withSeasonToggled(NO_FILTERS, 'fall'))).toBe(false);
    expect(isUnfiltered(withOccasionSelected(NO_FILTERS, 'casual'))).toBe(false);
  });
});

File: presentation/src/test/kotlin/com/wardrobapp/presentation/OutfitFiltersTest.kt
package com.wardrobapp.presentation

import com.wardrobapp.domain.Occasion
import com.wardrobapp.domain.Season
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

/**
 * The two chip rows above the outfit suggestions.
 *
 * They behave differently on purpose: seasons are a set, because a garment for
 * spring is often a garment for fall, and occasion is one choice, because an
 * outfit is for one thing at a time. The rule both rows share is that every state
 * can be got back out of -- there is no combination of taps that leaves someone
 * stuck with a filter they cannot clear.
 */
class OutfitFiltersTest {

    @Test
    fun `nothing chosen is the unfiltered state, and the any chips are the ones lit`() {
        val filters = OutfitFilters()

        assertTrue(filters.isUnfiltered)
        assertEquals(FilterChip<Season>(null, true), filters.seasonChips().first())
        assertEquals(FilterChip<Occasion>(null, true), filters.occasionChips().first())
        // One chip per season and occasion, plus the "any" chip in front.
        assertEquals(Season.entries.size + 1, filters.seasonChips().size)
        assertEquals(Occasion.entries.size + 1, filters.occasionChips().size)
    }

    @Test
    fun `tapping a season turns it on, and tapping it again turns it off`() {
        val one = OutfitFilters().withSeasonToggled(Season.SUMMER)
        assertEquals(listOf(Season.SUMMER), one.seasons)
        assertTrue(!one.isUnfiltered)

        val off = one.withSeasonToggled(Season.SUMMER)
        assertEquals(emptyList(), off.seasons)
        assertTrue(off.isUnfiltered, "emptying the row by tapping did not reach the unfiltered state")
    }

    @Test
    fun `seasons are kept in the app's own order, not the order they were tapped`() {
        // So the same choice always reads the same, and matches what the chips
        // show.
        val filters = OutfitFilters()
            .withSeasonToggled(Season.WINTER)
            .withSeasonToggled(Season.SPRING)

        assertEquals(Season.entries.filter { it in filters.seasons }, filters.seasons)
    }

    @Test
    fun `any clears the seasons rather than being one of them`() {
        val filters = OutfitFilters(seasons = listOf(Season.SUMMER, Season.WINTER))

        val cleared = filters.withSeasonToggled(null)

        assertEquals(emptyList(), cleared.seasons)
        assertEquals(FilterChip<Season>(null, true), cleared.seasonChips().first())
    }

    @Test
    fun `an occasion is one choice, and choosing another replaces it`() {
        val work = OutfitFilters().withOccasionSelected(Occasion.WORK)
        assertEquals(Occasion.WORK, work.occasion)

        val formal = work.withOccasionSelected(Occasion.FORMAL)
        assertEquals(Occasion.FORMAL, formal.occasion)
    }

    @Test
    fun `tapping the chosen occasion clears it, the same as tapping any`() {
        val work = OutfitFilters().withOccasionSelected(Occasion.WORK)

        assertEquals(null, work.withOccasionSelected(Occasion.WORK).occasion)
        assertEquals(null, work.withOccasionSelected(null).occasion)
    }

    @Test
    fun `the chips say what is on`() {
        val filters = OutfitFilters(seasons = listOf(Season.SUMMER), occasion = Occasion.WORK)

        val seasons = filters.seasonChips()
        assertEquals(FilterChip<Season>(null, false), seasons.first())
        assertEquals(listOf(Season.SUMMER), seasons.filter { it.active }.map { it.value })

        val occasions = filters.occasionChips()
        assertEquals(FilterChip<Occasion>(null, false), occasions.first())
        assertEquals(listOf(Occasion.WORK), occasions.filter { it.active }.map { it.value })
    }
}
